### ISSUE TITLE
Async methods in HttpChannel cause deadlocks when used in Task.WaitAll up in call stack

### ISSUE THREAD

**Issue by DotNetPart** [dotnet/wcf#2055]
Hi,

In example below if I do `Task.WaitAll( someObj.RequestAsync(channel, messafge) )` it causes deadlocks.
My initial suspicion is that's because inner async methods of HttpClientRequestChannel don't set ConfigureAwait(false). I'm aware that it's a really bad idea to do Task.WaitAll on async task and I'm trying to refactor into using await, but I'm still curious if there is any particular reason why async methods in this repo don't have ConfigureAwait(false).

```
private async Task<Message> RequestAsync(IRequestChannel requestChannel, Message message)
      {
         var taskCompletionSource = new TaskCompletionSource<Message>();
         try
         {
            requestChannel.BeginRequest(message, asyncResult => { }, null);
         }
         catch
         {
            taskCompletionSource.TrySetResult(default(Message));
            throw;
         }

         return await taskCompletionSource.Task.ConfigureAwait(false);
      }
```

**Comment by mconnew** [comment]
When using ConfigureAwait(false), it only does anything if the Task actually went async. If the Task doesn't go async, then you need to use ConfigureAwait(false) on the next Task that you intend to await. This also means you need to use ConfigureAwait(false) all the way down the call stack at every await occurrence. As a WCF operation is guaranteed to always go async at some point, we took a different approach. At every public entry point we detect if there's a non-default synchronization context set, and if there is, we immediately force the rest of the method to continue on a thread pool thread. This way we don't have the overhead of adding ConfigureAwait absolutely everywhere in our code (and there would be a really high likelihood of missing something somewhere which would be very intermittent and hard to track down) and just needs to be done in a small handful of places. An example of this is [here](https://github.com/dotnet/wcf/blob/master/src/System.Private.ServiceModel/src/System/ServiceModel/Channels/RequestChannel.cs#L258) in `RequestChannel.RequestAsyncInternal` which is called by RequestChannel.Request. Our functionality has been designed so that if you are "playing nice" and await-ing an async method, then everything works. If you want to wait synchronously on an operation, use the synchronous overload of the method. In other words, call RequestChannel.Request if you want to synchronously wait for the result and we will ensure the internal continuations get posted to the thread pool and won't deadlock. If you want to asynchronously wait, then use the async method. There should never be a need to synchronously wait on a WCF async operation as we provide synchronous overloads for everything.
Also, there's multiple issues with the code snippet you posted for calling APM methods and exposing a Task, there's a lot of edge cases that you miss. I'm not going to list them all out, but instead will [point to how it's done correctly](https://github.com/dotnet/wcf/blob/master/src/System.Private.ServiceModel/tests/Scenarios/Client/ChannelLayer/DuplexChannelShapeTests.4.0.0.cs#L107) using Task.Factory.FromAsync. Basically you can replace your code with this:
```csharp
private Task<Message> RequestAsync(IRequestChannel requestChannel, Message message)
{
    return Task.Factory.FromAsync((asyncCallback, o) => requestChannel.BeginRequest(requestMessage, asyncCallback, o), channel.EndSend, TaskCreationOptions.None);
}
```
You are doing something quite different though, you are running multiple async methods and so you don't want to block a thread for each one so aren't using the synchronous Request method. But then you are wanting to block waiting for the result of multiple requests instead of using await. You're kind of straddling both worlds. This means you need to take a little bit of responsibility for making sure the continuations don't deadlock but that is actually quite easy. If you have multiple of these calls that you wish to wait on completion of so want to use Task.WaitAll, then do the following:
```csharp
Task.Run(Task.WhenAll(someObj.RequestAsync(channel, message), someObj.RequestAsync(channel2, message2)).Wait();
```
You'll notice I'm not using actually using WaitAll, instead I'm using WhenAll which creates another Task which completes when all the child Task's complete. This allows asynchronously waiting for all your tasks to complete on another thread without blocking on two separate threads. Task.Run would normally return a Task<Task> in this scenario but automatically unwraps the inner Task so you can directly wait on the Task to complete. Task.Run executes the code on the thread pool so your local SynchronizationContext is ignored.  
We did things this way to avoid the performance overhead of calling ConfigureAwait hundreds of times for a single request and to ensure that our synchronous methods are guaranteed to never deadlock. Please let me know if this answers your question and resolves your problem.

**Comment by DotNetPart** [comment]
@mconnew Thank you. I have been able to refactor my code and get rid of Task.WaitAll having await on async task instead. However, I'm interested in getting deeper into details of your answer and make sure I understand it.

**Comment by mconnew** [comment]
@DotNetPart, if you have any questions, feel free to ask. There's a lot to unpack in my response :) I would still recommend switching to using Task.Factory.FromAsync if you haven't yet.

**Comment by mconnew** [comment]
Closing as working as designed and @DotNetPart is unblocked.
